frontends/concrete-python/concrete/fhe/mlir/context.py

```diff
@@ -2037,6 +2038,122 @@ def encrypt(self, resulting_type: ConversionType, x: Conversion) -> Conversion:
+        x: Conversion,
+        bit: Union[int, np.integer],
+    ) -> Conversion:
+        if bit >= x.original_bit_width:
+            return (
+                self.zeros(resulting_type)
+                if x.is_unsigned
+                else self.extract_bit(resulting_type, x, x.bit_width - 1)
+            )
+
+        for _ in range(bit):
+            lsb = self.lsb(x.type, x)
+            cleared = self.sub(x.type, x, lsb)
+            x = self.reinterpret(cleared, bit_width=(x.bit_width - 1))
+
+        result = self.to_signedness(self.lsb(x.type, x), of=resulting_type)
+
+        if result.bit_width > resulting_type.bit_width:
+            difference = result.bit_width - resulting_type.bit_width
+            shifter = self.constant(self.i(result.bit_width + 1), 2**difference)
+            shifted = self.mul(result.type, result, shifter)
+            result = self.reinterpret(shifted, bit_width=resulting_type.bit_width)
+
+        elif result.bit_width < resulting_type.bit_width:
+            shifter = self.constant(self.i(result.bit_width + 1), 2 ** (result.bit_width - 1))
+            shifted = self.mul(result.type, result, shifter)
+            reinterpreted = self.reinterpret(shifted, bit_width=1)
+            result = self.tlu(resulting_type, reinterpreted, [0, 1])
+
+        return result
+
+    def extract_bits(
+        self,
+        resulting_type: ConversionType,
+        x: Conversion,
+        bits: Union[int, np.integer, slice],
+    ) -> Conversion:
+        if x.is_clear:
+            highlights: Dict[Node, Union[str, List[str]]] = {
+                x.origin: "operand is clear",
+                self.converting: "but clear bit extraction is not supported",
+            }
+            self.error(highlights)
+
+        if isinstance(bits, (int, np.integer)):
+            return self.extract_bit(resulting_type, x, bits)
+
+        assert isinstance(bits, slice)
+
+        step = bits.step or 1
+
+        assert step != 0
+        if step < 0:
+            assert bits.start is not None
+
+        start = bits.start or MIN_EXTRACTABLE_BIT
+        stop = bits.stop or (MAX_EXTRACTABLE_BIT if step > 0 else MIN_EXTRACTABLE_BIT)
+
+        bits_and_their_positions = []
+        for position, bit in enumerate(range(start, stop, step)):
+            bits_and_their_positions.append((bit, position))
+
+        bits_and_their_positions = sorted(
+            bits_and_their_positions,
+            key=lambda bit_and_its_position: bit_and_its_position[0],
+        )
+
+        current_bit = 0
+        max_bit = x.original_bit_width
+
+        lsb: Optional[Conversion] = None
+        result: Optional[Conversion] = None
+
+        for index, (bit, position) in enumerate(bits_and_their_positions):
+            if bit >= max_bit and x.is_unsigned:
+                break
+
+            last = index == len(bits_and_their_positions) - 1
+            while bit != (current_bit - 1):
+                lsb = self.lsb(x.type, x)
```

Comment:
if last, you can spare the last lsb in many case, if the last bit position in x is suitable.
This is the case for ReLU, and multiplication for instance.
```python
if last and position == 0 (or 1 ?):
    lsb = self.reinterpret(x, ...)
```

Reply:
Last doesn't mean msb, it means it's the last requested bit. I think what you have in mind is:
```
                if bit == (max_bit - 1):
                    assert x.bit_width == 1
                    lsb = x
                else:
                    lsb = self.lsb(x.type, x)
```